File: Tracks/typescript/clock/clock.ts
export default class Clock {
  private static module = Object.freeze({
    hour: 24,
    minute: 60
  });

  private readonly hours: number;
  private readonly minutes: number;

  private static toModulated = (
    input: number,
    type: keyof typeof Clock.module
  ): number => {
    if (input > 0) return input;
    const module = Clock.module[type];
    return input + (Math.abs(Math.floor(input / module)) + 1) * module;
  };

  constructor(hours: number, minutes: number = 0) {
    this.minutes =
      Clock.toModulated(minutes, "minute") % Clock.module["minute"];
    this.hours =
      Clock.toModulated(
        hours + Math.floor(minutes / Clock.module["minute"]),
        "hour"
      ) % Clock.module["hour"];
  }

  plus = (minutes: number): Clock =>
    new Clock(this.hours, this.minutes + minutes);

  minus = (minutes: number): Clock => this.plus(-1 * minutes);

  equals = (other: Clock): boolean =>
    this.hours === other.hours && this.minutes === other.minutes;

  private static toTwoDigits = (input: number): string =>
    input.toFixed(0).padStart(2, "0");

  toString = (): string =>
    `${Clock.toTwoDigits(this.hours)}:${Clock.toTwoDigits(this.minutes)}`;
}

File: Tracks/csharp/clock/Clock.cs
using System;
using System.Diagnostics.CodeAnalysis;

public class Clock : IEquatable<Clock>
{
    public bool Equals([AllowNull] Clock other) => (hours, minutes) == (other.hours, other.minutes);

    private readonly int hours;
    private readonly int minutes;
    private const int MINUTES_IN_ONE_HOUR = 60;
    private const int HOURS_IN_ONE_DAY = 24;

    private (int extra, int modulated) ToPositiveModule(int number, int module)
    {
        double timesOverModule = (double)Math.Abs(number) / module;
        (int upper, int lower) = ((int)Math.Ceiling(timesOverModule), (int)Math.Floor(timesOverModule));
        return number < 0
            ? (-upper, number + upper * module)
            : (lower, number - lower * module);
    }

    public Clock(int hours, int minutes)
    {
        (int extraHours, int newMinutes) = ToPositiveModule(minutes, MINUTES_IN_ONE_HOUR);
        (this.hours, this.minutes) = (ToPositiveModule(hours + extraHours, HOURS_IN_ONE_DAY).modulated, newMinutes);
    }

    public Clock Add(int minutesToAdd) => new Clock(hours, minutes + minutesToAdd);


    public Clock Subtract(int minutesToSubtract) => Add(-minutesToSubtract);

    public override string ToString() => $"{hours:D2}:{minutes:D2}";
}
